src/wallet/rpcwallet.cpp

```diff
@@ -3080,50 +3087,119 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
     if (!fExperimentalOrchardWallet) {
         throw JSONRPCError(RPC_WALLET_ENCRYPTION_FAILED, "Error: the Orchard wallet experimental extensions are disabled.");
     }
-    int64_t account = params[0].get_int64();
-    if (account < 0 || account >= ZCASH_LEGACY_ACCOUNT) {
+
+    LOCK(pwalletMain->cs_wallet);
+
+    int64_t accountInt = params[0].get_int64();
+    if (accountInt < 0 || accountInt >= ZCASH_LEGACY_ACCOUNT) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid account number, must be 0 <= account <= (2^31)-2.");
     }
-    // TODO: Check that the account is known to the wallet.
-    std::vector<uint8_t> parsed_diversifier_index;
+    libzcash::AccountId account = accountInt;
+
+    std::optional<libzcash::diversifier_index_t> j = std::nullopt;
     if (params.size() >= 2) {
+        // TODO: Handle '*'
         if (params[1].getType() != UniValue::VNUM) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid diversifier index, must be an unsigned integer.");
```

Comment:
@LarryRuane implemented this parameter in #5300 as a JSON number, which prevents us from using `*` as the default here (as earlier verification logic will reject it as being a string input to a numberic field). We could make it a JSON string argument instead (which I think would then prevent users passing numbers here). However, I think I'd prefer that we use a diversifier index of `-1` to indicate "any", since diversifier indices will always be unsigned.

Reply:
I think that instead of using `-1`, we should change the order of arguments such that the list of pools comes first, and then is followed by the diversifier index as the last argument. Then, the default is provided just by omitting the argument, and I expect choosing pools to be more common and important an operation than specifying a diversifier index manually anyway, and the empty list is a good value to pass for "use all the default pools."

Reply:
I feel a bit uncomfortable using the empty list for "yeah, default is good" rather than "no receivers please", but we can never have an empty receiver set, so I guess it's as meaningful as `-1` would be (in that they use some unoccupiable part of the domain space as a sentinel). I'll make this change.